src/lib/viewports/viewport.js

```diff
@@ -245,30 +201,72 @@ export default class Viewport {
       width: this.width,
       height: this.height,
       scale: this.scale
-    },
-
-      // Subclass can add additional params
-      // TODO - Fragile: better to make base Viewport class aware of all params
-      this._getParams()
-    );
+    });
 
     return matrices;
   }
 
   getDistanceScales() {
-    return {
-      pixelsPerMeter: [1, 1, 1],
-      metersPerPixel: [1, 1, 1],
-      pixelsPerDegree: [1, 1, 1],
-      degreesPerPixel: [1, 1, 1]
-    };
+    return this.distanceScales;
+  }
+
+  getCameraPosition() {
+    return this.cameraPosition;
   }
 
   // INTERNAL METHODS
 
-  // Can be subclassed to add additional fields to `getMatrices`
-  _getParams() {
-    return {};
+  _initMatrices() {
+    // Note: As usual, matrix operations should be applied in "reverse" order
+    // since vectors will be multiplied in from the right during transformation
+    const vpm = createMat4();
+    mat4_multiply(vpm, vpm, this.projectionMatrix);
+    mat4_multiply(vpm, vpm, this.viewMatrix);
+    this.viewProjectionMatrix = vpm;
+
+    // Calculate inverse view matrix
+    this.viewMatrixInverse = mat4_invert([], this.viewMatrix) || this.viewMatrix;
+
+    // Read the translation from the inverse view matrix
+    this.cameraPosition = [
+      this.viewMatrixInverse[12],
+      this.viewMatrixInverse[13],
+      this.viewMatrixInverse[14]
+    ];
+
+    this.cameraDirection = [
+      this.viewMatrix[2],
+      this.viewMatrix[6],
+      this.viewMatrix[10]
+    ];
+
+    this.cameraUp = [
+      this.viewMatrix[1],
+      this.viewMatrix[5],
+      this.viewMatrix[9]
+    ];
+
+    /*
+     * Builds matrices that converts preprojected lngLats to screen pixels
+     * and vice versa.
+     * Note: Currently returns bottom-left coordinates!
+     * Note: Starts with the GL projection matrix and adds steps to the
+     *       scale and translate that matrix onto the window.
+     * Note: WebGL controls clip space to screen projection with gl.viewport
+     *       and does not need this step.
+     */
+
+    // matrix for conversion from location to screen coordinates
+    const m = createMat4();
+    mat4_scale(m, m, [this.width / 2, -this.height / 2, 1]);
+    mat4_translate(m, m, [1, -1, 0]);
```

Comment:
The translation is to invert the y-axis. You know, y axis is different in the map space and screen space. This set of translation is copied from Mapbox and they do have a weird translation system for viewport. I can explain to you in details later